apps-rendering/src/pagination.ts

```diff
@@ -0,0 +1,146 @@
+import { map, none, some, withDefault } from '@guardian/types';
+import type { Option } from '@guardian/types';
+import { pipe } from 'lib';
+import type { LiveBlock } from 'liveBlock';
+
+export type PageReference = {
+	blocks: LiveBlock[];
+	pageNumber: number;
+	suffix: string;
+};
+
+export type Pagination = {
+	newest: Option<string>;
+	newer: Option<string>;
+	oldest: Option<string>;
+	older: Option<string>;
+	numberOfPages: number;
+};
+
+export type LiveBlogPagedBlocks = {
+	currentPage: PageReference;
+	pagination: Pagination;
+};
+
+const navContainerId = 'liveblog-navigation';
+
+const getOldestPage = (
+	pages: LiveBlock[][],
+	pageNumber: number,
+): Option<string> => {
+	if (pageNumber >= pages.length) return none;
+
+	const blocks = pages[pages.length - 1];
+	return some(`?page=with:block-${blocks[0].id}#${navContainerId}`);
+};
+
+const getOlderPage = (
+	pages: LiveBlock[][],
+	pageNumber: number,
+): Option<string> => {
+	if (pageNumber < pages.length) {
+		const blocks = pages[pageNumber];
+		return some(`?page=with:block-${blocks[0].id}#${navContainerId}`);
+	}
+
+	return none;
+};
+
+const getNewerPage = (
+	pages: LiveBlock[][],
+	pageNumber: number,
+): Option<string> => {
+	if (pageNumber > 2) {
+		const blocks = pages[pageNumber - 2];
+		return some(`?page=with:block-${blocks[0].id}#${navContainerId}`);
+	}
+
+	if (pageNumber === 2) {
+		return some(`?`);
+	}
+
+	return none;
+};
+
+const getNewestPage = (
+	pages: LiveBlock[][],
+	pageNumber: number,
+): Option<string> => {
+	if (pageNumber > 1) {
+		return some(`?`);
+	}
+
+	return none;
+};
+
+const getFirstPageLength = (pageSize: number, blocks: LiveBlock[]): number => {
+	const remainder = blocks.length % pageSize;
+	return pageSize + remainder;
+};
+
+const getPages = (
```

Comment:
This function does a lot of the groundwork for us: it's returning us an array of liveblock arrays that represent the pages of the liveblock. When we looked at the CAPI data model we were confident that the liveblocks are returned in order (i.e. newest first) so there wasn't any need to sort the liveblocks before creating pages.

Reply:
Actually, good spot from @marjisound - we possibly do need to consider ordering the blocks (https://github.com/guardian/frontend/blob/2996c4465a09ed108c64e7d12c0b35d042dbcce7/common/app/model/liveblog/BodyBlock.scala#L17)

Reply:
I spoke with the CAPI team about ordering of blocks. The CAPI team does not apply any ordering on the blocks, they send in the same order that Composer provides them (I suppose this is where we could possibly introduce out-of-order blocks). 

For now, I'm not going to introduce sorting of blocks - instead I'm going to try and evaluate liveblog blocks to see if I can find examples of any out-of-order blocks. I'll also check with the developers who originally introduced sorting into frontend (https://github.com/guardian/frontend/pull/13566).

Reply:
We have ~28k articles that have tag `tone/minutebyminute` (`https://content.guardianapis.com/search?tag=(tone/minutebyminute)&to-date=2022-01-12&show-blocks=body`). I sampled 1000 of these articles that have this tag:
- I performed a simple sort on the liveblog blocks (order descending by firstPublishedDate)
- I compared the ordered blocks to the blocks we receive from CAPI

All 1000 blogs I sampled had the expected ordered. I'll expand the test to see if I can find any examples where the blocks are not ordered as expected.